platform/linux-generic/odp_ml.c

```diff
@@ -0,0 +1,2673 @@
+		*result = result_local;
+
+	return ret;
+}
+
+static inline int check_compl_param(const odp_ml_compl_param_t *compl_param,
+				    uint32_t max_compl_id, odp_bool_t is_load)
+{
+	odp_ml_config_t *config = &_odp_ml_glb->ml_config;
+
+	switch (compl_param->mode) {
+	case ODP_ML_COMPL_MODE_POLL:
+		if (is_load && !(config->load_mode_mask & ODP_ML_COMPL_MODE_POLL)) {
+			_ODP_ERR("Poll mode loading/unloading is not configured\n");
+			return -1;
+		}
+
+		if (!is_load && !(config->run_mode_mask & ODP_ML_COMPL_MODE_POLL)) {
+			_ODP_ERR("Poll mode run is not configured\n");
+			return -1;
+		}
+
+		if (compl_param->compl_id > max_compl_id) {
+			_ODP_ERR("Bad compl_id: %u, exceeding model max completion id %u\n",
+				 compl_param->compl_id, max_compl_id);
+			return -1;
+		}
+		break;
+	case ODP_ML_COMPL_MODE_EVENT:
+		if (is_load && !(config->load_mode_mask & ODP_ML_COMPL_MODE_EVENT)) {
+			_ODP_ERR("Event mode loading/unloading is not configured\n");
+			return -1;
+		}
+
+		if (!is_load && !(config->run_mode_mask & ODP_ML_COMPL_MODE_EVENT)) {
+			_ODP_ERR("Event mode run is not configured\n");
+			return -1;
+		}
+
+		if (compl_param->event == ODP_EVENT_INVALID ||
+		    compl_param->queue == ODP_QUEUE_INVALID) {
+			_ODP_ERR("Bad event or queue\n");
+			return -1;
+		}
+
+		if (odp_event_type(compl_param->event) != ODP_EVENT_ML_COMPL) {
+			_ODP_ERR("Bad completion event type\n");
+			return -1;
+		}
+		break;
+	default:
+		/* Including ODP_ML_COMPL_MODE_SYNC, which is not supported by
+		 * asynchrous functions (e.g. *_start()) either.
+		 */
+		_ODP_ERR("Invalid completion mode %u\n", compl_param->mode);
+		return -1;
+	}
+
+	return 0;
+}
+
+int odp_ml_model_load_start(odp_ml_model_t model, const odp_ml_compl_param_t *compl_param)
+{
+	int ret;
+	ml_model_t *mdl = ml_model_from_handle(model);
+
+	if (odp_unlikely(model == ODP_ML_MODEL_INVALID)) {
+		_ODP_ERR("Bad model handle\n");
+		return -1;
+	}
+
+	if (odp_unlikely(check_compl_param(compl_param, mdl->max_compl_id, true)))
+		return -1;
+
+	if (compl_param->mode == ODP_ML_COMPL_MODE_POLL)
+		odp_atomic_store_rel_u32(&mdl->compl_status[compl_param->compl_id], 0);
+
+	ret = odp_ml_model_load(model, NULL);
+
+	if (odp_unlikely(ret))
+		return -1;
+
+	/* Send a completion event to the given queue */
+	if (compl_param->mode == ODP_ML_COMPL_MODE_EVENT) {
+		odp_ml_load_result_t *result;
+		odp_buffer_t buf = (odp_buffer_t)(uintptr_t)compl_param->event;
+
+		_odp_buffer_subtype_set(buf, ODP_EVENT_ML_COMPL_LOAD);
+
+		result = odp_buffer_addr(buf);
+		result->error_code = 0;
+		result->user_ptr = compl_param->user_ptr;
+
+		if (odp_unlikely(odp_queue_enq(compl_param->queue, compl_param->event))) {
+			_ODP_ERR("Completion event enqueue failed %" PRIu64 "\n",
+				 odp_queue_to_u64(compl_param->queue));
+			return -1;
+		}
+
+		return 0;
+	}
+
+	odp_atomic_store_rel_u32(&mdl->compl_status[compl_param->compl_id], 1);
+	mdl->result[compl_param->compl_id].user_ptr = compl_param->user_ptr;
```

Comment:
When atomic rel/acq is used for status flag, this result write should be before status flag write.